### ISSUE TITLE
Fast concatenation of bit vectors?

### ISSUE THREAD

**Issue by Ryex** [myrrlyn/bitvec#63]
I'm writing an.. interesting project where I'm basically making a canonicalize tree to represent a 2d bit space. this is so I can store results for nodes of the tree and their children that are the same value

it is written such that the leaf nodes contain small u8 backed bit-vectors

on occasion, I need to extract the full bit space form the tree. my current implementation into allocating a BitVec that has enough capacity for the full 2d space (avoiding reallocation) and then descend through the tree using `extend` to fill in the vector.

my current benchmark has this process at 4-5ms for my problem space (be aware I'm compiling to wasm so this won't quite be native speed)

This is fast, but it also appears to be a bottleneck. I figure that iterating over bits is relatively inefficient when I could be using the backing elements but I'm can't think of a clean way to speed up this process.

any suggestions?

**Comment by myrrlyn** [comment]
Unfortunately, there is not a general solution to accelerating the copy operation used by `.extend()`. An API like this:

```rust
impl<O, T> BitVec<O, T>
where O: BitOrder, T: BitStore {
  pub fn concat(&mut self, other: &BitSlice<O, T>);
}
```

would have an opportunity to attempt element-wise copying, with the requirement that `self` end at an element boundary and `other` begin at one.

I've had this problem in the back of my mind for a while but I've not been able to come up with something that can adequately solve all the constraints here.

If you are able to guarantee that your leaf vectors always start and end at element boundaries, then you may have better performance by using `.as_slice()` to accumulate the backing elements, and then reconstructing a bit-vector after accumulation. If you cannot guarantee this, then the performance is, unfortunately, restricted to bit-by-bit.

**Comment by Ryex** [comment]
I eventually decided I do exactly that, My solution is to limit my tree sizes so that I could guarantee that my leaf nodes fill element boundaries and then use a `Vec<u8>` to accumulate elements using `.as_slice()` and then construct a bitvector from the `Vec<u8>`

profiler is now saying that I spend about next to no time in iterators on bitvecs and 22% in the actual computations. I think it's about as good as I'm gonna get.

now my main performance hit is coming from the hash operations at 11.4%
```
Percent│     _ZN4core4sync6atomic11atomic_load17h82cd3ce7feb3ab53E():
       │
       │     #[inline]
       │     unsafe fn atomic_load<T: Copy>(dst: *const T, order: Ordering) -> T {
       │     match order {
       │     Acquire => intrinsics::atomic_load_acq(dst),
       │     Relaxed => intrinsics::atomic_load_relaxed(dst),
 16.74 │       movzbl (%rdi),%ebx
``` 

and PartialEq at 5%
```
Percent│     _ZN4core4sync6atomic11atomic_load17h82cd3ce7feb3ab53E():
       │
       │     #[inline]
       │     unsafe fn atomic_load<T: Copy>(dst: *const T, order: Ordering) -> T {
       │     match order {
       │     Acquire => intrinsics::atomic_load_acq(dst),
       │     Relaxed => intrinsics::atomic_load_relaxed(dst),
 21.26 │       movzbl (%rdx),%ebx
```

any thoughts on reducing the number of atomic load operations needed there?

**Comment by Ryex** [comment]
Actually, to respond the that later part, I changed my Hash and PartialEq impls use `.as_slice()` because I don't have to worry about edge elements,

Flat 15-20% speedup across my benchmarks. this makes perfect sense and I'm not sure what I didn't think of it before.

**Comment by myrrlyn** [comment]
If you have a slice produced by the use of `.split_at_mut`, it will be tainted with atomic operations across its whole length. You can use `.bit_domain()` or `.domain()` to access `&BitSlice`s or `&[T]` slices, respectively, that restrict atomic access to only the addresses that strictly require it, and use non-atomic accesses for the full extent of the addresses that do not.

This behavior *should* be present in the `0.17` series, and is definitely present in the `0.18` series; I just need to finish work on that and get it released.

**Comment by myrrlyn** [comment]
I released 0.18. I am going to mark this issue as closed, since it appears that between your comments and my notes, the crate is now as capable as it can be and you've found a pattern that works for you. Please re&ouml;pen if you are still seeing needless locks.
